File: subArrSumK.py
#Time Complexity :  O(n)
#Space Complexity : O(n)
#Did this code successfully run on Leetcode : Yes
#Any problem you faced while coding this : No

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        res = 0
        curSum = 0
        d = {0:1}
        
        for num in nums:
            curSum = curSum + num
            
            #if difference is already in dict keys then update res
            #this if condition is checking for continuous sum
            if curSum-k in d.keys():
                res = res + d[curSum-k]
            
            #if currSum not in the dict keys then add curSum as key
            #and set the value to be 1
            if curSum not in d.keys():
                d[curSum] = 1
            #if the key is already. present in dict then increment by 1
            else:
                d[curSum] = d[curSum] + 1
        
        return res

File: SubArrSumK.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : Understanding the logic itself. It is hard problem for me.

/*
1. Calculate Running Sum
2. See Running-K has happened before then add those many increments in to Count
3. Then add Running Sum in to Map
4. Return Count
*/

//Running Sum Pattern

import java.util.HashMap;
import java.util.Map;

public class SubArrSumK {
    public int subarraySum(int[] nums, int k) {
        int count=0;
        if(nums == null || nums.length==0) return count;
        int runningSum=0;
        Map<Integer,Integer> map = new HashMap<>();
        map.put(0, 1); //dummy entry for running 0 happens 1 time: To catch earlier SubArray with sumk

//        for (int num : nums) {
        for(int i =0; i< nums.length; i++) {
            runningSum += nums[i];

            if(map.containsKey(runningSum - k)){
                count+= map.get(runningSum - k);
            }

            if(map.containsKey(runningSum)){
                map.put(runningSum, map.get(runningSum)+1);
            }else{
                map.put(runningSum, 1);
            }
        }
        return count;
    }

    public static void main(String[] args) {
        SubArrSumK obj = new SubArrSumK();
//        System.out.println("Count =  "+ obj.subarraySum(new int[]{1,1,1}, 2));
        System.out.println("Count =  "+ obj.subarraySum(new int[]{1}, 0));
    }
}
